intervals.go

```
@@ -0,0 +1,64 @@
+package epoch
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Interval struct {
+	Unit  Unit
+	Value float64
+}
+
+func ParseInterval(interval string) (*Interval, error) {
+	interval = strings.TrimSpace(interval)
+	if len(interval) < 2 {
+		return nil, fmt.Errorf("interval string is too short")
+	}
+	var value float64
+	var unitStr string
+	n, err := fmt.Sscanf(interval, "%f%s", &value, &unitStr)
+	if n != 2 || err != nil {
+		return nil, fmt.Errorf("failed to parse interval string: %s", err)
+	}
+	unit := ListAvailableUnits().Get(unitStr)
+	if unit == nil {
+		return nil, fmt.Errorf("unknown unit %q", unitStr)
+	}
+	return &Interval{Unit: *unit, Value: value}, nil
+}
+
+func MustParseInterval(interval string) *Interval {
+	i, err := ParseInterval(interval)
+	if err != nil {
+		panic(err)
+	}
+	return i
+}
+
+func (i Interval) String() string {
+	return strconv.FormatFloat(i.Value, 'f', -1, 64) + i.Unit.Short
+}
+
+func (i Interval) Duration() time.Duration {
```

Do you see issues with this function?
Are all options of this switch OK?

---

The Duration() method will work as expected, but there are a couple of issues with the current implementation.
First of all, it assumes that there are exactly 30 days in a month and 365 days in a year, which is not always the case because the number of days in a month can vary between 28 to 31 and a leap year has 366 days.
Secondly, The time.Duration type is based on nanoseconds but duration calculation you do is in terms of seconds, but It should be in terms of nanoseconds.

It is not a big problem for small intervals, as the error will be rounded to about 1% for intervals of one month and about 0.3% for intervals of one year, but it will be significant for large intervals.
If the accuracy is a concern, the time package should be used to do more precise calculations. It is possible to use time.Time and time.Duration to calculate time intervals and durations with greater accuracy.
Please let me know if you need any more help.

---

Great, right. I want `interval.Duration` to not return results if they are not precise

---

Alright, in that case, you can simply remove the case statements for `UnitMonth` and `UnitYear` in the `Duration()` function, since they are not accurate.

And you can also remove the d variable and directly return the `time.Duration`
And you can add a default case statement that will panic if you encounter an unexpected unit value .
Here is the updated code `interval.Duration()`